plugin/environment.go

```diff
@@ -114,13 +114,14 @@ func (env *Environment) GetPluginState(id string) int {
 		return model.PluginStateNotRunning
 	}
 
-	return *rp.(*registeredPlugin).State
+	return rp.(*registeredPlugin).State
 }
 
 // SetPluginState sets the current state of a plugin (disabled, running, or error)
 func (env *Environment) SetPluginState(id string, state int) {
 	if rp, ok := env.registeredPlugins.Load(id); ok {
-		*rp.(*registeredPlugin).State = state
+		rp.(*registeredPlugin).State = state
+		env.registeredPlugins.Store(id, rp)
```

Comment:
Since we're storing pointers to a `registeredPlugin` in the map, it seems like this `Store` is redundant -- aren't we already mutating on the line above? (Even with `State` as a value type?)

Is this a potential cause for another data race? Should we consider storing value types of registeredPointers as well?

Reply:
I believe you are right @lieut-data. While this PR fixes the race for the `.State` access, the race due to the pointer to `*registeredPlugin` is still there. This was actually uncovered by @streamer45 here https://mattermost.atlassian.net/browse/MM-21019.

I think the right fix to this is to increase the lock scope from the `sync.Map` to use a general mutex and guard the access to the registeredPlugin also. So `registeredPlugins` becomes a simple map, and we stick in a mutex to the `Environment` struct for methods which mess with the state.

This would actually be a good change, because I believe `sync.Map` was not the perfect choice for this. From the [documentation](https://golang.org/pkg/sync/#Map), it mentions that `sync.Map` is optimized for cases when a key is only written once and read many times across several goroutines, or when there are disjoint sets of keys. Both of which is inapplicable here. So I think this is actually good for the codebase.

How about I send another PR for this. so that we keep this PR just for the State variable access ?

Reply:
> it mentions that sync.Map is optimized for cases when a key is only written once and read many times across several goroutines

I actually think `sync.Map` is perfect for this use case -- the vast majority of cases are reads, and we only write infrequently: on startup, or when a user adds/removes a plugin. Both of these pale in comparison to the read load.

> I think the right fix to this is to increase the lock scope from the sync.Map to use a general mutex and guard the access to the registeredPlugin also.

We intentionally switched to `sync.Map` to avoid re-entrancy issues with `sync.RWutex`: https://github.com/mattermost/mattermost-server/pull/9167.

I /think/ we can eliminate the race altogether if we just abandon the need to maintain a pointer to a `registeredPlugin` and just store values in the map instead. That, combined with your change to have `State` be a simple value, and it seems like we can keep `sync.Map` and avoid the data races. Thoughts?

Reply:
> I actually think sync.Map is perfect for this use case -- the vast majority of cases are reads,
Ooh okay, that's nice :+1: 

> 
I /think/ we can eliminate the race altogether if we just abandon the need to maintain a pointer to a registeredPlugin and just store values in the map instead. 
Sure, we can go ahead with that.

Reply:
@agnivade, I'm not seeing the changes discussed above to abandon the pointer to a `registeredPlugin` and just store values in the map instead. From what I can see, there's still a race here in that we're directly mutating the `State` value of an object that might also be being read. 

Should we switch to just storing values in the `sync.Map` instead? It's possible I'm misunderstanding something, though!